### PR TITLE
Bugfix/objects 1054

### PR DESCRIPTION
**Pull request by andreassiegel**
### What changes were proposed in this pull request?

This adds a `NotFoundController` that is capable of returning _404 Not Found_ responses at the gateway if a forwarding route is configured:

```
catch-all:
  path: /rest/**
  url: forward:/notFound
  stripPrefix: false
```

The `PreAuthorizationFilter` is changed to only filter if the request should not be forwarded to `/notFound`.

In addition, the error controller is updated to return improved error messages like

```
{
  "timestamp": 1476968316440,
  "status": 503,
  "error": "Service Unavailable",
  "message": "Forwarding error (Load balancer does not have available server for client: profiles-ui-welcome)",
  "path": "/rest/things/bob"
}
```

It particularly differentiates between Timeouts and Client Exceptions. These will return _504 Gateway Timeout_ or _503 Service Unavailable_ respectively. All other responses will remain _500 Internal Server Error_.
All responses will contain a body now.
### How is this patch documented?

Code
### How was this patch tested?
- manually using Postman
- additional unit tests
#### Depends On

Nothing, but the _404_ response will only work if the route is present in the configuration.

### CODE REVIEW

**Review by ghost — APPROVED** [review]
LGTM

**Comment by wfschneider-rfid** [comment]
Test Failures?

**Comment by wfschneider-rfid** [comment]
ok.

File: src/main/java/net/smartcosmos/cluster/gateway/resource/GatewayErrorController.java [hunk 1]
```diff
@@ -27,51 +37,189 @@
...
+                                       + "Status code: '%s', Error: '%s'\n"
+                                       + "Cause: %s\n"
+                                       + "Root cause: %s",
+                                       route,
+                                       serviceId,
+                                       requestUri,
+                                       httpStatus,
+                                       errorMessage,
+                                       errorException != null ? errorException.toString() : "No exception available in context",
+                                       rootCause != null ? rootCause.toString() : "N/A");
```
Comment (jbauerrfid) [inline src/main/java/net/smartcosmos/cluster/gateway/resource/GatewayErrorController.java]: Better use a `StringBuilder` here to concatenate this amount of text elements.

**Pull request merged**